### ISSUE TITLE
0.12 Plan - feedback welcome

### ISSUE THREAD

**Issue by yyx990803** [vuejs/Discussion#158]
Hi Vue.js users, here is a list of changes I'm planning for the 0.12 release. The goal is to remove some cruft in the API and make things more consistent across the board. The two new features have already been implemented in the `dev` branch, and I will soon start doing 0.12 beta releases so people can try it out. The breaking changes are pretty much just ideas that I'd like to gather feedback for - in particular, let me know how hard do you think it would be for you to migrate your existing app if these changes went live.

### New Features
- async components. Allows defining a component as a function that asynchronous resolves the component definition. This would enable easy lazy-loading mechanisms.
- one-way paramAttributes/props syntax (`prop="{{*onetime}}"`)

### Breaking Changes
- `paramAttributes` renamed to `props`.
- Intent to remove: `v-partial` and `{{>partial}}` syntax. The partial concept is a legacy from string-based templates, and isn't particularly useful when everything should just be components.
- Intent to remove: `v-with` syntax.
  - It allows the user to add arbitrary data fields to a child component, which may conflict with the child's default data. At the same time, the child needs to be more explicit about what it expects from the parent.
  - It's also confusing to have two ways of doing the same thing. Everything should just be `props` (previously `paramAttributes`).
  - The name `v-with` isn't super precise for what it does now: passing data down to a child. 
- Intent to remove: `v-component` syntax.
  - Similar to `v-with`, it doesn't make much sense to have two ways of doing the same thing. Every component should just be a custom tag.
  - For dynamic components, use the special `<dynamic-component type="{{view}}"></dynamic-component>` syntax. `"dynamic-component"` is a reserved component name.
  - It should no longer be required to have a hyphen in your component name in order to use it as a custom tag.
- Intent to remove: `replace` option. A component always replaces its placeholder node.
- Intent to remove: `inherit` option. Prefer explicitly pass down props.

**Comment by holic** [comment]
:+1: 

I'm not sure on the proposed `inherit` changes. I was working on an app where I tried to make use of `paramAttributes` to keep my components isolated without tight coupling, but ran into issues. I'll see if I can remember or recreate them so I can mention them here.

**Comment by thelinuxlich** [comment]
Agree with everything minus removing the inherit option.

**Comment by taylorotwell** [comment]
Looks pretty good overall!

**Comment by kazupon** [comment]
:+1: 

Among them, I agree especially with removing the `v-with` directive. because it breaking the child component data.

**Comment by nadirabid** [comment]
:+1:

Vue's simple and flexible API is paramount. To be honest, I haven't personally found need for async but I can see it being useful. 

Looking forward to 0.12!

**Comment by cheapsteak** [comment]
Breaking v-with would certainly be a pain to refactor were we to upgrade, but I think it's the right way to go. :+1:

**Comment by OEvgeny** [comment]
Looks pretty good. Except maybe inheritance. But I think with proper ``props`` implementation we will not need to inherit all data. Anyway ``inherit`` breaks component system, I think.

As I thought one of the partials purposes was to decrease number of complete Vue instances and replace it with something lightweight. Am I wrong? To be honest, didn't use it overall, but it looks useful.

Also Vue documentation is great, but sometimes I need it to be more reach. Any plans for it?

Waiting for  0.12 too!

**Comment by ayamflow** [comment]
:+1: pretty good to see Vue move closer to "real" web components.

**Comment by pechorin** [comment]
:+1: 

Removing `v-with` and `v-component` is great simplification. Also `props` look's more like react and this is also nice idea.

**Comment by crswll** [comment]
:+1: This all sounds great to me.

**Comment by azamat-sharapov** [comment]
This is great!!!!

Does "one-way props" have performance advantages? Or it's just for better isolation/protection of prop?

**Comment by yyx990803** [comment]
@azamat-sharapov performance will be largely the same, it's for better isolation. In some cases you don't want the component to be able to mutate parent state.

**Comment by azamat-sharapov** [comment]
@yyx990803 thanks. Personally for my projects, if I decide to switch to 0.12, I will have to get rid of `v-component`, I have been using it since it's not always easy to come up with good name for component without dash. So now dash is not necessary and it's helpful. 
Never used partials and always avoid use of `inherit` since it looks like bad practice to me, so +1 for removing it. And I still have `v-with` in some places, though I lately got used to paramAttributes and think it's better, so +1 for removing `v-with` too, even though it requires refactoring for many.

**Comment by young-steveo** [comment]
@yyx990803  I seem to be in the minority when it comes to `v-component`;  I'm a fan of the `<div v-component="{{ view }}"></div>` syntax.  I never use custom tags in my projects (no strong opinion against custom tags, I just prefer the aesthetics of attributes).  This isn't the end of the world, and I love using Vue, so I'll migrate to `<dynamic-component type="{{ view }}"></dynamic-component>` if this change is set in stone.

Love the idea of async components! :+1:

**Comment by crswll** [comment]
@young-steveo Agreed, but I think it's just us being stubborn. It's definitely going to be the norm in the future.

**Comment by yyx990803** [comment]
Update:
- Simplified `<dynamic-component type="{{view}}">` to just `<component type="{{view}}">`.
- `replace` option will be preserved, due to it requiring some significant refactoring for some existing apps. For those of you would like to have it default to `true`, you can do so by setting `Vue.options.replace = true`.
- `inherit` option will be preserved, although it should be treated as "only use if you know what you are doing"...

**Comment by steffans** [comment]
:+1: 

What about naming dynamic components `<v-component type="{{ view }}"><v-component>`, it's shorter and it matches the current attribute namespace style.

**Comment by holic** [comment]
> `replace` option will be preserved, due to it requiring some significant refactoring for some existing apps. 

Curious what these use-cases are that would require major refactoring?

**Comment by 285858315** [comment]
async components 这个真的很赞！

**Comment by 285858315** [comment]
async components 这个真的很赞！

**Comment by sjoerdvisscher** [comment]
If `<component type="{{view}}">` is only used for dynamic content, you could make type a reactive directive, instead of a literal one: `<component type="view">`

**Comment by azamat-sharapov** [comment]
Why is it `type` anyway? how about `name` or `variable` or maybe `source` ? can't understand meaning of `type`..

**Comment by yyx990803** [comment]
@sjoerdvisscher you can still do `<component type="component-id">` and it resolves to the `component-id` literal value. `type` here is similar to a prop and the braces make it explicit that it is bound to a dynamic value.

**Comment by taylorotwell** [comment]
Agree that name "type" is not very intuitive.

**Comment by yyx990803** [comment]
What would be better though? I don't think `name`, `variable` or `source` is any better than `type`.

**Comment by crswll** [comment]
It tells you what `type` of component it is just like `<input type="password">` tells you it's an `<input>` of type `password`. I think it makes sense.

**Comment by RangerMauve** [comment]
I agree that type makes sense. However, might I suggest borrowing `is` from the web components spec? Or would that just make things more confusing like having a dash in the name did.

**Comment by yyx990803** [comment]
@RangerMauve `is` sounds like a plausible idea, and it resembles real Web Components.

**Comment by kzima** [comment]
+1 for async components and lazy-loading mechanisms. However, I wonder how would that play with webpack?

**Comment by yyx990803** [comment]
@kzima like this:

``` js
components: {
  'my-async-component': function (resolve) {
    require.ensure(['./my-async-component.js'], function () {
      resolve(require('./my-async-component.js'))
    })
  }
}
```

**Comment by azamat-sharapov** [comment]
@yyx990803 no passing `require` parameter?

**Comment by yyx990803** [comment]
@azamat-sharapov fixed

**Comment by ayamflow** [comment]
@yyx990803 the only thing that feels "unatural" is the `<component type="..">` syntax.
I saw something about a `<{{component}} />`in the ractive repo issues but I don't think it's functional. I guess there's not a lot of alternatives tho.

**Comment by azamat-sharapov** [comment]
`type` sounded wierd for me in the beginning too, but when @crswll mentioned that it's like `<input type="something">`, it made sense. But for newcomers, it will probably be confusing, especially if they are not much into frontend stuff. For example Taylor Otwell is more server-side-code guru and thinks that `type` is not very intuitive, whereas Evan - Javascript/HTML5 ninja prefers `type`. Although `is` sounds very easy as mentioned here and probably is better choice overall :)

**Comment by thelinuxlich** [comment]
+1 for "is"

**Comment by OEvgeny** [comment]
For me ``type`` sounds good.

**Comment by kazupon** [comment]
+1 for `is`

**Comment by Nirazul** [comment]
+1 for `is`

I wonder why `v-partial` was removed and what the altermative would look like. I especially liked it for really flexible components that have multiple appearances (f. ex. selector-modules that could assume the form of a dropdown, a list or something other). The appearance could be decided by choosing a type, which was very convenient!

**Comment by yyx990803** [comment]
Update: 
- Dynamic component syntax now uses `is` since people like it and it also brings the concept closer to the real web components spec.

The change is included in [0.12.0-beta3](https://github.com/yyx990803/vue/releases/tag/0.12.0-beta3).

**Comment by OEvgeny** [comment]
Wonder why people love ``is`` so much. At first  ``is`` is a verb. Now we have (only one?) attribute what is verb instead of noun. Anyway, it's not very important. I'm only trying to understand why  :smile:

**Comment by azamat-sharapov** [comment]
It's not about they love it :) It began, because `type` was little hard to figure out, for most people. As noted earlier, `is` it's more near to real web components specs (http://webcomponents.org/articles/introduction-to-custom-elements/). Also, I personally think vue should be simple to understand for everybody, so they focus on code, read through code faster. `is` is more readable. 

`<components type="{{currentView}}">`

vs.

`<components is="{{currentView}}">`

**Comment by OEvgeny** [comment]
Wow, thanks for the link. Now I see ``is`` is a good choice.

**Comment by ayamflow** [comment]
@yyx990803 By the way, have you considered moving some of the Vue internal to external modules/repos? Since the library is getting a bit bigger that it originally was, it might be useful for code reuse, debugging, extending and maintainability to not have everything under the same repo. Thoughts?

I'm thinking with the observation and templating in mind, since this is something that will come soon to the browser with ES6/7. Since you are trying to follow as much as possible the Webcomponents spec, maybe using `Object.observe()` specs as well for instance? It would make it easy to switch to the native API when supporte, thus lightening the compiled script.

For instance, I've been using Ampersand lately and I love that everything is a module that you can change. A good example is their observation module: [ampersand-state](https://github.com/ampersandjs/ampersand-state).

**Comment by yyx990803** [comment]
#### Update: Filter Syntax Change

Some thoughts on filter arguments syntax improvements:

- Currently filter arguments are treated as plain strings, and can be enclosed in quotes.
- In `filterBy` and `orderBy`, in order to differentiate between dynamic key and static key, the quotes in arguments are preserved.
- A problem with this though, is that quotes are also preserved for other filters such as `pluralize`, and all custom filters as well.
- This is also inconsistent, because in `filterBy` and `orderBy` you need quotes to indicate a plain string, but in other filters you don't.

#### Proposed fix (breaking):

All arguments to filters, if not in quotes, are treated as dynamic paths (and the argument will be auto-retrieved from the vm when the filter function is called); only arguments enclosed in quotes are passed in as plain strings.

This means the usage of some existing filters will have to change:

```
<a v-on="keyup: onKeyUp | key 'enter'"></a>
{{ items.length | pluralize 'item' }}
```

**Comment by RangerMauve** [comment]
Yeah, having quotes around arguments that should be strings makes sense. It feels less "magical", too.

**Comment by yyx990803** [comment]
@ayamflow I've been thinking about this, and my general opinion is that there is no "right level of modularization". People have different preferences on how fine-grained modularization should be: some people like monolithic solutions; some people like picking their own view-layer, persistence-layer and router; some people like going even lower level and hand-pick every single component (e.g. AmpersandJS). But being finer-grained is not strictly better than the other solutions, because the choice is really about how much flexibility you want VS. how much time you are willing to invest in picking your own components.

From a library's perspective, it is about what the core value the library is designed to provide: the whole system consisted of these modules, or the reusability of the modules out of the system context. The latter is what AmpersandJS is all about, but is not what defines Vue as a project.

Finally, I can assure you breaking Vue into smaller modules would increase the cost of maintenance rather decreasing it. So given the limited energy I can put into Vue, I don't think I'll do that in the foreseeable future. (Maybe when `Object.observe` reaches candidate stage)

**Comment by holic** [comment]
:+1: For the proposed filter syntax change.

**Comment by ayamflow** [comment]
@yyx990803 yes this totally makes sense. Thanks for the detailed answer.
I think the true benefit would be for all things related to specs so at some point it can be dropped for the native implementation. But since none of that is ready yet... :)

**Comment by kazupon** [comment]
:+1: For filter syntax change

**Comment by steffans** [comment]
@yyx990803 I like the new filter syntax :+1: Support for object definitions would be also nice, see pull request: https://github.com/yyx990803/vue/pull/834

**Comment by fullfs** [comment]
Great changes!
Though I prety much never use components as custom elements. Used to v-component syntax.

P.S. Never had any problems making async components. ```this.$options.components[name] = Vue.extend({...})``` for win :D

**Comment by azamat-sharapov** [comment]
@fullfs do "file/replace in all files using some regex". This might save some time for you or anybody who is refactoring: https://regex101.com/r/yH5iU8/2 (achtung! that's quick regex, don't take it too serious).

**Comment by fullfs** [comment]
@azamat-sharapov aye, but simple replace won't be enough in my case. Have to do some logic corretions too.

**Comment by fullfs** [comment]
Oh and I use partials a lot. It's very useful sometimes. Definitely don't want to convert them all into components...

**Comment by young-steveo** [comment]
@fullfs It seems you have built a pretty large body of work around a `pre-v1.0.0` library.  That can sometimes get pretty painful as the API stabilizes.

**Comment by fullfs** [comment]
@young-steveo true. And I'm aware it. It was a risk to begin with. But I can give some useful feedback :D So I'd like partials to be kept :)

**Comment by yyx990803** [comment]
@fullfs if you really want partials that bad, we can maybe ship it as a standalone directive separately.

**Comment by mark-hahn** [comment]
> @OEvgeny said ...
> Wonder why people love is so much. At first is is a verb.

The word "is" is an indicative which is nothing like a noun or verb.  An indicative indicates what something else is.  So its usage is perfect here.

**Comment by kazupon** [comment]
In async components feature, 
when the undefined component found, if we can resolve lazy-loading with hook function, I think that is very useful.

this like:

```js
var Vue = require('Vue')
var _ = Vue.util

new Vue({
  ...
  
  components: {
    // with object
    component1: {
      ...
    },
    // with url path
    component2: 'components/component2.js',
    ...
  },
  ...
  
  // lazy-loading hook function
  loading: function (raw, resolve) {
    if (_.isPlainObject(raw) { // component plain object
      resolve(raw)
    } else if (typeof raw === 'string') { // component url path
      // lazy-loading with require.js
      require([raw], resolve)
      
      /* lazy-loading with webpack
      require.ensure([raw], function (require) {
        resolve(require(raw))
      })
      */
    }
  },
  ...
}).$mount('#app')
...
```

what do you think about it ?

**Comment by yyx990803** [comment]
@kazupon one problem with this is that module bundler like Browserify and Webpack rely on seeing the literal string of module id (e.g. `require('my-component')`) in order to resolve dependencies between modules. When you `require(dynamicVar)`, they won't be able to tell which module you are trying to require.

This should work with requirejs though. You can make a helper function:
``` js
function async (path) {
  return function (resolve) {
    require([path], resolve)
  }
}

Vue.component(async('my-component.js'))
```

**Comment by fullfs** [comment]
@yyx990803 by the way, please notice error handling while getting async components. There is a problem with the case while using webpack.

**Comment by yyx990803** [comment]
@fullfs can you elaborate?

**Comment by fullfs** [comment]
@yyx990803 
Ok, for example we've got an async component require. In webpack we go like this first:

```
var component = require('bundle?lazy&name=component!My/Async/Component');
```

Then we try to pull the component:
```
component(function (file) {
    resolve(file);
});
```

The thing is there is no way we can catch an error while using webpack async require (I've tried a plugin for the case, but it didn't work properly too) if server didn't respond (server may go down or something). The callback just won't fire. Then my app just will be freezed.

So there is a need for some way to catch the error (timeout?) and say something like "oops we couldn't get the component, so lets fall back".

It could be a ```reject``` callback, for example:

```
components: {
  'my-async-component': function (resolve, reject) {
    var timeout = setTimeout(function () {
      reject('timeout');
    }, 3000);
    require(['./my-async-component'], function (file) {
      clearTimeout(timeout)
      resolve(file)
    })
  }
}
```

**Comment by kazupon** [comment]
@yyx990803 
I see !!
I will try to use helper function. :)

**Comment by yyx990803** [comment]
@fullfs it might be helpful to have some warning, but if you failed to resolve an async component, what would you expect as the fallback? There's no component to render, so the app really can't do anything. Calling the callback when the component resolution has failed will just result in an error.

**Comment by fullfs** [comment]
@yyx990803 When async component's failed to resolve I want a way for my app to react, that's it. Resolving another component is a good thing. It's actually the fallback I'd like to have (for some reason didn't think I could do that :D).
Thanks :)

**Comment by yyx990803** [comment]
Some thoughts on improving DOM event and Vue event handling syntax: https://github.com/yyx990803/vue/issues/722#issuecomment-109849898

**Comment by Nirazul** [comment]
I don't like the proposal concerning `v-on`. I imagine that it can get very sluggish when you want to handle multiple dom-events on the same element. 
For me it's no improvement to switch out a custom tag with an also namespaced tag which is different for every event that exists and is to come in the future.
I like the possibility to pass a dynamic object to `v-on`, rather than writing custom tags for everything.

As for `v-events` I've recently had some struggle to differenciate between using `v-events` as a Directive and the `events` property. Maybe these two could be made clearer.

**Comment by fullfs** [comment]
@yyx990803 
By the way, talking about version 0.12. Is there any hope to have 0.11 -> 0.12 migration plugin? :) 
I would have really hard time with the migration if no, after all. There are 3 things I'd like to have there: ```v-component```, ```v-with``` and ```partials```.

**Comment by thelinuxlich** [comment]
I don't see any benefits for the new event syntax :(

**Comment by yyx990803** [comment]
Looks like most people want to stick with `v-` directives, we'll keep things as is for now then.

**Comment by Nirazul** [comment]
`v-callbacks` sounds alright, maybe `v-listeners` would be clearer for the purpose. It sort of "listens" for events happening in its child.

**Comment by yyx990803** [comment]
@Nirazul I've decided to just remove it. You can now just pass down parent methods as props and let children call them.

**Comment by federicodionisi** [comment]
@yyx990803
As @fullfs suggested, would be really nice to have a migration plugin from 0.11.0 to 0.12.0

All new features inside 0.12.0 are amazing but most of them are breaking changes and for those who are using Vue in production will be painful to update it.

**Comment by cjsaylor** [comment]
@fullfs @federicodionisi You guys should collaborate and make one. I don't think it's necessary for a library author of pre-release software to also have to contend with writing migration scripts.

**Comment by fullfs** [comment]
@yyx990803 Can I ask you for solution about partials? Now it's gone and they need to be replaced by something.
I usully use partials as something similar to template includes to make my templates smaller.
For example, I'm selling apartments and I've got a page with about one with several sets of props.
There are a lot of props there, but all of them are just a part of single data object. And so I've got several partials with sets of the params for each view case.
What should I do now? The simplest solution is to make them into components with ```inherit: true```, but it looks overhead to me, I don't want to create extra VMs.
Is there any recommended way?

**Comment by yyx990803** [comment]
@fullfs does making `v-html` auto compile its content sound like a solution?

**Comment by fullfs** [comment]
@yyx990803 I'm afraid it's not. I won't be able to use ```v-on``` and etc. there, right?

**Comment by Nirazul** [comment]
I have the same feeling about `v-partial`. It was the excellent solution for a component that can take multiple shapes that differ only in markup and css, but not in functionality. Now I'm forced to use `v-if`, `v-html` or make a new component.

**Comment by yyx990803** [comment]
Yeah, ever since 0.12 I've seen people run into cases where partial just makes more sense. I'll consider adding it back in 0.12.2.

**Comment by fullfs** [comment]
@yyx990803 That would be absolutely great! Thanks!
